xrspatial/zonal.py

```diff
@@ -657,14 +654,155 @@ def _crosstab_dask(
     return crosstab_df
 
 
+def _find_cats(values, cat_ids, nodata_values):
+    if len(values.shape) == 2:
+        # 2D case
+        unique_cats = np.unique(values.data[
+            np.isfinite(values.data) & (values.data != nodata_values)
+        ])
+        if cat_ids is None:
+            cat_ids = unique_cats
+        # remove cats that does not exist in `values` raster
+        cat_ids = [c for c in cat_ids if c in unique_cats]
+    else:
+        # 3D case
+        unique_cats = values[values.dims[0]].data
+        if cat_ids is None:
+            cat_ids = unique_cats
+        # remove cats that does not exist in `values` raster
+        cat_ids = [c for c in cat_ids if c in unique_cats]
+    return unique_cats, cat_ids
+
+
+def _sort_values(sorted_indices, values, cats):
+    if len(values.shape) == 2:
+        # 2D case
+        result = values.flatten()[sorted_indices]
```

Comment:
Would it be better to use `ravel` here, which only copies if necessary, rather than `flatten` which always copies?